### ISSUE TITLE
double file lock causes conda-build to hang (conda-build 2.0.0beta2)

### ISSUE THREAD

**Issue by ericdill** [conda/conda-build#1250]
There is a pathway where conda-build attempts to enter the [filelock context manager] (https://github.com/conda/conda-build/blob/master/conda_build/build.py#L405) twice and thus hangs.

Attempting to build the example recipe from conda-forge/staged recipes with the current tip of the conda-build master branch (2.0.0beta2-9-gae8a97b) results in this behavior on my system.

The following stack trace is from `ctrl+c`-ing while conda-build appears to be hung

```
$ conda build example/
Removing old build environment
BUILD START: simplejson-3.8.2-py35_0
    (actual version deferred until further download or env creation)
updating index in: /home/eric/miniconda/conda-bld/linux-64
updating index in: /home/eric/miniconda/conda-bld/noarch
Using Anaconda API: https://api.anaconda.org
WARNING: Could not detect installed version of conda-build

The following NEW packages will be INSTALLED:

    ca-certificates: 2016.8.2-3      conda-forge
    certifi:         2016.8.2-py35_0 conda-forge
    ncurses:         5.9-8           conda-forge
    openssl:         1.0.2h-2        conda-forge
    pip:             8.1.2-py35_0    conda-forge
    python:          3.5.2-2         conda-forge
    readline:        6.2-0           conda-forge
    setuptools:      25.1.6-py35_0   conda-forge
    sqlite:          3.13.0-1        conda-forge
    tk:              8.5.19-0        conda-forge
    wheel:           0.29.0-py35_0   conda-forge
    xz:              5.2.2-0         conda-forge
    zlib:            1.2.8-3         conda-forge

WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:Build prefix failed with prefix length 255.
WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:Error was: 
WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:ERROR: placeholder '/opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho' too short in: conda-forge::ncurses-5.9-8

WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:One or more of your package dependencies needs to be rebuilt with a longer prefix length.
WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:Falling back to legacy prefix length of 80 characters.
WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:Your package will not install into prefixes > 80 characters.
updating index in: /home/eric/miniconda/conda-bld/linux-64
updating index in: /home/eric/miniconda/conda-bld/noarch
WARNING: Could not detect installed version of conda-build
^CTraceback (most recent call last):
  File "/home/eric/miniconda/lib/python3.5/site-packages/conda/install.py", line 1076, in link
    update_prefix(join(prefix, f), prefix, placeholder, mode)
  File "/home/eric/miniconda/lib/python3.5/site-packages/conda/install.py", line 425, in update_prefix
    data = replace_prefix(mode, data, placeholder, new_prefix)
  File "/home/eric/miniconda/lib/python3.5/site-packages/conda/install.py", line 407, in replace_prefix
    data = binary_replace(data, placeholder.encode('utf-8'), new_prefix.encode('utf-8'))
  File "/home/eric/miniconda/lib/python3.5/site-packages/conda/install.py", line 380, in binary_replace
    res = pat.sub(replace, data)
  File "/home/eric/miniconda/lib/python3.5/site-packages/conda/install.py", line 377, in replace
    raise PaddingError(a, b, padding)
conda.install.PaddingError: (b'/opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho', b'/home/eric/miniconda/conda-bld/_1471970593628/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl', -175)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/eric/dev/conda/conda-build/conda_build/build.py", line 413, in create_env
    plan.execute_actions(actions, index, verbose=config.debug)
  File "/home/eric/miniconda/lib/python3.5/site-packages/conda/plan.py", line 599, in execute_actions
    inst.execute_instructions(plan, index, verbose)
  File "/home/eric/miniconda/lib/python3.5/site-packages/conda/instructions.py", line 135, in execute_instructions
    cmd(state, arg)
  File "/home/eric/miniconda/lib/python3.5/site-packages/conda/instructions.py", line 78, in LINK_CMD
    link(state['prefix'], dist, lt, index=state['index'])
  File "/home/eric/miniconda/lib/python3.5/site-packages/conda/install.py", line 1079, in link
    (placeholder, dist))
SystemExit: ERROR: placeholder '/opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho' too short in: conda-forge::ncurses-5.9-8


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/eric/miniconda/bin/conda-build", line 9, in <module>
    load_entry_point('conda-build', 'console_scripts', 'conda-build')()
  File "/home/eric/dev/conda/conda-build/conda_build/cli/main_build.py", line 256, in main
    execute(sys.argv[1:])
  File "/home/eric/dev/conda/conda-build/conda_build/cli/main_build.py", line 234, in execute
    already_built=None, config=config)
  File "/home/eric/dev/conda/conda-build/conda_build/api.py", line 77, in build
    need_source_download=need_source_download, config=config)
  File "/home/eric/dev/conda/conda-build/conda_build/build.py", line 896, in build_tree
    config=config)
  File "/home/eric/dev/conda/conda-build/conda_build/build.py", line 539, in build
    create_env(config.build_prefix, specs, config=config)
  File "/home/eric/dev/conda/conda-build/conda_build/build.py", line 434, in create_env
    clear_cache=clear_cache)
  File "/home/eric/dev/conda/conda-build/conda_build/build.py", line 406, in create_env
    timeout=config.timeout):
  File "/home/eric/miniconda/lib/python3.5/site-packages/filelock.py", line 297, in __enter__
    self.acquire()
  File "/home/eric/miniconda/lib/python3.5/site-packages/filelock.py", line 245, in acquire
    time.sleep(poll_intervall)
KeyboardInterrupt
```

If I let it run to completion, this is the full console output:

```
$ conda build example/
Removing old build environment
BUILD START: simplejson-3.8.2-py35_0
    (actual version deferred until further download or env creation)
updating index in: /home/eric/miniconda/conda-bld/linux-64
updating index in: /home/eric/miniconda/conda-bld/noarch
Using Anaconda API: https://api.anaconda.org
WARNING: Could not detect installed version of conda-build

The following NEW packages will be INSTALLED:

    ca-certificates: 2016.8.2-3      conda-forge
    certifi:         2016.8.2-py35_0 conda-forge
    ncurses:         5.9-8           conda-forge
    openssl:         1.0.2h-2        conda-forge
    pip:             8.1.2-py35_0    conda-forge
    python:          3.5.2-2         conda-forge
    readline:        6.2-0           conda-forge
    setuptools:      25.1.6-py35_0   conda-forge
    sqlite:          3.13.0-1        conda-forge
    tk:              8.5.19-0        conda-forge
    wheel:           0.29.0-py35_0   conda-forge
    xz:              5.2.2-0         conda-forge
    zlib:            1.2.8-3         conda-forge

WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:Build prefix failed with prefix length 255.
WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:Error was: 
WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:ERROR: placeholder '/opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho' too short in: conda-forge::ncurses-5.9-8

WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:One or more of your package dependencies needs to be rebuilt with a longer prefix length.
WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:Falling back to legacy prefix length of 80 characters.
WARNING:/home/eric/dev/conda/conda-build/conda_build/build.py:Your package will not install into prefixes > 80 characters.
updating index in: /home/eric/miniconda/conda-bld/linux-64
updating index in: /home/eric/miniconda/conda-bld/noarch
WARNING: Could not detect installed version of conda-build
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda-build/issues

Include the output of the command 'conda info' in your report.
```

and, of course:
```
$ conda build --version
conda-build 2.0.0beta2+9.gae8a97b
```

**Comment by msarahan** [comment]
Thanks for the report.  Will fix ASAP.  Your testing of master is much appreciated.

**Comment by ericdill** [comment]
happy to help!

**Comment by msarahan** [comment]
I think that I have fixed this in #1252 - I changed the with to manual lock management.  I don't think the with part lets go of the lock in case of recursion.

**Comment by ericdill** [comment]
This is fixed in 2.0.0beta3 which includes #1252. Thanks for the quick fix @msarahan .  I'll run conda build master again to try and uncover more issues now that this is resolved.
